File: app/src/main - 副本/java/com/zmcursor/mydaliy/Model/ImageLoader.java
package com.zmcursor.mydaliy.Model;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Bitmap;

import com.zmcursor.mydaliy.AppConfig;
import com.zmcursor.mydaliy.Model.Bean.BaseStory;
import com.zmcursor.mydaliy.Model.Bean.StoriesOfDate;
import com.zmcursor.mydaliy.Model.Bean.Story;
import com.zmcursor.mydaliy.Model.Bean.TopStory;
import com.zmcursor.mydaliy.Model.Storage.DataStorage;

import java.util.List;

/**
 * Created by ZMcursor on 2018/5/30 0030.
 */

public class ImageLoader {

    private static final int maxSavedDate = 7;

    private Context context;
    private DateList dateList;

    public ImageLoader(Context context) {
        this.context = context;

        SharedPreferences sp = context.getSharedPreferences(AppConfig.save_data, Context.MODE_PRIVATE);
        dateList = new DateList(sp.getString(AppConfig.date_list_key, null));
    }

    public void saveDateList() {
        SharedPreferences sp = context.getSharedPreferences(AppConfig.save_data, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putString(AppConfig.date_list_key, dateList.toString());
        editor.apply();
    }

    public void getTopImages(List<TopStory> topStories) {
        DataStorage.initTopImage(context);
        for (TopStory topStory : topStories) {
            if (topStory.getImageBitmap() == null) {
                Bitmap image = DataStorage.getImage(context, null, topStory.id);
                if (image == null) downloadImage(null, topStory, true);
            }
        }
        DataStorage.deleteOldTopImage(context);
    }

    public void getImages(StoriesOfDate storiesOfDate, String beforeDate) {
        String old = dateList.add(storiesOfDate.getDate(), beforeDate);
        boolean saveImage = true;
        boolean dateSaved = false;
        if ("0".equals(old)) saveImage = false;
        else if ("1".equals(old)) dateSaved = true;
        else if (old != null) DataStorage.deleteImageOfDate(context, old);

        for (Story story : storiesOfDate.getStories()) {
            if (story.getImageBitmap() == null) {
                Bitmap image = null;
                if (dateSaved)
                    image = DataStorage.getImage(context, storiesOfDate.getDate(), story.id);
                if (image == null) downloadImage(storiesOfDate.getDate(), story, saveImage);
            }
        }
    }

    private void downloadImage(String date, BaseStory story, boolean isSave) {
        DataLoader.download(false, story.getImageUrl(), data -> {
            if (data != null) {
                Bitmap image = (Bitmap) data;
                story.setImageBitmap(image);
                if (isSave) DataStorage.saveImage(context, image, date, story.id);
            }
        });
    }

    private class DateList {
        private String[] dateList;

        DateList(String list) {
            if (list != null) dateList = splitDate(list, '#');
            else dateList = new String[maxSavedDate];
        }

        String add(String date, String beforeDate) {
            int pos = 0;
            if (beforeDate != null) {
                pos = have(beforeDate) + 1;
                if (pos >= dateList.length) return "0";
            }
            if (!date.equals(dateList[pos])) {
                return add(pos, date);
            }
            return "1";
        }

        private int have(String date) {
            int pos = 0;
            while (pos < dateList.length && !date.equals(dateList[pos])) pos++;
            return pos;
        }

        private String add(int pos, String date) {
            String old;
            int i = dateList.length - 1;
            old = dateList[i];
            while (i > pos) {
                dateList[i] = dateList[--i];
            }
            dateList[i] = date;
            return old;
        }

        private String[] splitDate(String list, char regex) {
            String[] dates = new String[maxSavedDate];
            for (short i = 0, r = 0, l = 0; r < list.length(); r++) {
                if (list.charAt(r) == regex) {
                    dates[i] = list.substring(l, r);
                    l = ++r;
                    i++;
                }
            }
            return dates;
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            for (String s : dateList) {
                if (s == null) break;
                sb.append(s);
                sb.append('#');
            }
            return sb.toString();
        }
    }
}

File: app/src/main/java/com/zmcursor/mydaliy/Model/ImageLoader.kt
package com.zmcursor.mydaliy.Model

import android.content.Context
import android.graphics.Bitmap
import com.zmcursor.mydaliy.AppConfig
import com.zmcursor.mydaliy.Model.Bean.BaseStory
import com.zmcursor.mydaliy.Model.Bean.StoriesOfDate
import com.zmcursor.mydaliy.Model.Bean.TopStory
import com.zmcursor.mydaliy.Model.Storage.DataStorage
import com.zmcursor.mydaliy.Net.OnDataGetListener

/**
 * Created by ZMcursor on 2018/5/30 0030.
 */

class ImageLoader(private val context: Context) {
    private val dateList: DateList

    init {
        val sp = context.getSharedPreferences(AppConfig.save_data, Context.MODE_PRIVATE)
        dateList = DateList(sp.getString(AppConfig.date_list_key, null))
    }

    fun saveDateList() {
        val sp = context.getSharedPreferences(AppConfig.save_data, Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.putString(AppConfig.date_list_key, dateList.toString())
        editor.apply()
    }

    fun getTopImages(topStories: List<TopStory>) {
        DataStorage.initTopImage(context)
        for (topStory in topStories) {
            if (topStory.imageBitmap == null) {
                val image = DataStorage.getImage(context, null, topStory.id)
                if (image == null) downloadImage(null, topStory, true)
            }
        }
        DataStorage.deleteOldTopImage(context)
    }

    fun getImages(storiesOfDate: StoriesOfDate, beforeDate: String) {
        val old = dateList.add(storiesOfDate.date, beforeDate)
        var saveImage = true
        var dateSaved = false
        when {
            "0" == old -> saveImage = false
            "1" == old -> dateSaved = true
            old != null -> DataStorage.deleteImageOfDate(context, old)
        }
//        if (storiesOfDate.stories == null) return
        for (story in storiesOfDate.stories!!) {
            if (story.imageBitmap == null) {
                var image: Bitmap? = null
                if (dateSaved)
                    image = DataStorage.getImage(context, storiesOfDate.date, story.id)
                if (image == null) downloadImage(storiesOfDate.date, story, saveImage)
            }
        }
    }

    private fun downloadImage(date: String?, story: BaseStory, isSave: Boolean) {
        DataLoader.download(story.imageUrl!!, object : OnDataGetListener<Bitmap> {
            override fun onDataGet(data: Bitmap) {
                story.imageBitmap = data
                if (isSave) DataStorage.saveImage(context, data, date!!, story.id)
            }
        })
    }

    private inner class DateList internal constructor(list: String?) {
        private var dateList: Array<String>? = null

        init {
            if (list != null)
                dateList = splitDate(list, '#')
            else
                dateList = arrayOfNulls(maxSavedDate)
        }

        internal fun add(date: String, beforeDate: String?): String? {
            var pos = 0
            if (beforeDate != null) {
                pos = have(beforeDate) + 1
                if (pos >= dateList!!.size) return "0"
            }
            return if (date != dateList!![pos]) {
                add(pos, date)
            } else "1"
        }

        private fun have(date: String): Int {
            var pos = 0
            while (pos < dateList!!.size && date != dateList!![pos]) pos++
            return pos
        }

        private fun add(pos: Int, date: String): String {
            val old: String
            var i = dateList!!.size - 1
            old = dateList!![i]
            while (i > pos) {
                dateList[i] = dateList!![--i]
            }
            dateList[i] = date
            return old
        }

        private fun splitDate(list: String, regex: Char): Array<String> {
            val dates = arrayOfNulls<String>(maxSavedDate)
            var i: Short = 0
            var r: Short = 0
            var l: Short = 0
            while (r < list.length) {
                if (list.get(r) == regex) {
                    dates[i] = list.substring(l.toInt(), r.toInt())
                    l = ++r
                    i++
                }
                r++
            }
            return dates
        }

        override fun toString(): String {
            val sb = StringBuilder()
            for (s in dateList!!) {
                if (s == null) break
                sb.append(s)
                sb.append('#')
            }
            return sb.toString()
        }
    }

    companion object {

        private val maxSavedDate = 7
    }
}
